integrations/access/msteams/app.go

```diff
@@ -187,27 +210,52 @@ func (a *App) initBot(ctx context.Context) error {
 	return nil
 }
 
-// newWatcherJob creates WatcherJob
-func (a *App) newWatcherJob() (lib.ServiceJob, error) {
-	return watcherjob.NewJob(
+// run starts the main process
+func (a *App) run(ctx context.Context) error {
+
+	process := lib.MustGetProcess(ctx)
+
+	watchKinds := []types.WatchKind{
+		{Kind: types.KindAccessRequest},
+		{Kind: types.KindAccessMonitoringRule},
+	}
+	acceptedWatchKinds := make([]string, 0, len(watchKinds))
+	watcherJob, err := watcherjob.NewJobWithConfirmedWatchKinds(
 		a.apiClient,
 		watcherjob.Config{
-			Watch: types.Watch{
-				Kinds: []types.WatchKind{{Kind: types.KindAccessRequest}},
-			},
+			Watch:            types.Watch{Kinds: watchKinds, AllowPartialSuccess: true},
 			EventFuncTimeout: handlerTimeout,
 		},
 		a.onWatcherEvent,
+		func(ws types.WatchStatus) {
+			for _, watchKind := range ws.GetKinds() {
+				acceptedWatchKinds = append(acceptedWatchKinds, watchKind.Kind)
+			}
+		},
 	)
-}
-
-// run starts the main process
-func (a *App) run(ctx context.Context) error {
-	ok, err := a.watcherJob.WaitReady(ctx)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
+	process.SpawnCriticalJob(watcherJob)
+
+	ok, err := watcherJob.WaitReady(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(acceptedWatchKinds) == 0 {
+		return trace.BadParameter("failed to initialize watcher for all the required resources: %+v",
+			watchKinds)
+	}
+	// Check if KindAccessMonitoringRule resources are being watched,
+	// the role the plugin is running as may not have access.
+	if slices.Contains(acceptedWatchKinds, types.KindAccessMonitoringRule) {
+		if err := a.accessMonitoringRules.InitAccessMonitoringRulesCache(ctx); err != nil {
+			return trace.Wrap(err, "initializing Access Monitoring Rule cache")
+		}
+	}
+	a.watcherJob = watcherJob
+	a.watcherJob.SetReady(ok)
```

Comment:
This code contains two races:
- we start processing AR events while the AMR cache is not inited
- we start processing AMR events while the AMR cache is not inited (cache should be inited when receiving the watcher INIT event)

This will lead to two issues:
- ARs processed early after plugin startup are not sent to the correct recipients
- AMRs created/updated early after plugin startup are ignored by the plugin

As those issues also exist in the common AMR plugin implementation we might decide to merge the PR anyway and fix those later.

Reply:
Are you suggesting queuing up incoming ARs and AMRs for processing until the init is done?

Reply:
For the AR vs AMR race I'm not sure yet, I would tend to create the AMR watcher first, then based on the result, setup AR watching. I'm worrdied this approach could lead out-of-order events but I'm not even sure Teleport protects against them in the first place with a dual-listener. I'll ask @tigrato and @espadolini if we have strong ordering guarantees in the auth event stream.

For the AMR init race, I would edit the `watcherjob` logic to make it run the cache init login on the `INIT` event. This is the proper way of initializing a cache in Teleport, but the watcherjob lib currently swallows the init event because it was not designed to maintain a local cache but only receive ARs.

Reply:
There's no protection against out of order events even with a single listener: watcher events are guaranteed to be delivered in order only with respects to each individual item's events, with no guarantee of ordering across items.

FWIW the `integrations/lib/watcherjob.job` seems to take full advantage of that, sequentially applying events for the same kind and name but processing events for different resources concurrently.

Reply:
And yeah, any replica fed by a watcher should open the watcher, wait for the `OpInit` event which is guaranteed to be the first event yielded by the watcher, then fetch the current state, then apply events in order.

Reply:
So plan for now is merge this then create a sperate issue and PR for the fixes to the access monitoring rule cache?
@hugoShaka

Reply:
Yes, all other plugins already have broken watchers so one more won't be an issue.